eth/beacon/db/chain.py

```diff
@@ -0,0 +1,450 @@
+
+            score = 0 if is_genesis else cls._get_score(db, first_block.parent_hash)
+
+        for block in blocks:
+            db.set(
+                block.hash,
+                rlp.encode(block),
+            )
+
+            # FIXME: It's a stub before we implement fork choice rule
+            score += block.slot_number
+
+            db.set(
+                SchemaV1.make_block_hash_to_score_lookup_key(block.hash),
+                rlp.encode(score, sedes=rlp.sedes.big_endian_int),
+            )
+
+        try:
+            previous_canonical_head = cls._get_canonical_head(db).hash
+            head_score = cls._get_score(db, previous_canonical_head)
+        except CanonicalHeadNotFound:
+            (
+                new_canonical_blocks,
+                old_canonical_blocks
+            ) = cls._set_as_canonical_chain_head(db, block.hash)
+        else:
+            if score > head_score:
+                (
+                    new_canonical_blocks,
+                    old_canonical_blocks
+                ) = cls._set_as_canonical_chain_head(db, block.hash)
+            else:
+                new_canonical_blocks = tuple()
+                old_canonical_blocks = tuple()
+
+        return new_canonical_blocks, old_canonical_blocks
+
+    @classmethod
+    def _set_as_canonical_chain_head(
+            cls, db: BaseDB,
+            block_hash: Hash32) -> Tuple[Tuple[BaseBeaconBlock, ...], Tuple[BaseBeaconBlock, ...]]:
+        """
+        Sets the canonical chain HEAD to the block as specified by the
+        given block hash.
+
+        :return: a tuple of the blocks that are newly in the canonical chain, and the blocks that
+            are no longer in the canonical chain
+
+        NOTE: copied from ``eth.db.header.HeaderDB._set_as_canonical_chain_head``,
+        and changed "header" to "block".
+        NOTE: modified a lot.
+        """
+        try:
+            block = cls._get_block_by_hash(db, block_hash)
+        except BlockNotFound:
+            raise ValueError(
+                "Cannot use unknown block hash as canonical head: {}".format(block_hash)
+            )
+
+        new_canonical_blocks = tuple(reversed(cls._find_new_ancestors(db, block)))
+        old_canonical_blocks = []
+
+        for block in new_canonical_blocks:
+            try:
+                old_canonical_hash = cls._get_canonical_block_hash(db, block.slot_number)
+            except BlockNotFound:
+                # no old_canonical block, and no more possible
+                break
+            else:
+                old_canonical_block = cls._get_block_by_hash(db, old_canonical_hash)
+                old_canonical_blocks.append(old_canonical_block)
+
+        for block in new_canonical_blocks:
+            cls._add_block_slot_to_hash_lookup(db, block)
+
+        db.set(SchemaV1.make_canonical_head_hash_lookup_key(), block.hash)
+
+        return new_canonical_blocks, tuple(old_canonical_blocks)
+
+    @classmethod
+    @to_tuple
```

Comment:
I think the two approaches are equivalent but I think generally `classmethod` belongs *closer* to the function definition and `to_tuple` should be above it.

Reply:
I tried:
```python
@to_tuple
@classmethod
def _find_new_ancestors(...)
```
then got error: `TypeError: 'classmethod' object is not callable`. It looks like we have to go through `to_tuple` and then `@classmethod` when calling it.